### PR TITLE
Add Unit Tests

### PR DESCRIPTION
**Pull request by Defelo**
## Description
Added unit tests for TDD with 100% coverage

## Related Issue
#9 

## How Has This Been Tested?
tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: app/mock/mock.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+from unittest.mock import MagicMock
+
+MicroService: MagicMock = MagicMock()
```
Comment (use-to) [inline app/mock/mock.py]: You can leave the typing out, because this is explicit.

File: app/tests/test_app.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        self.assertEqual(mock.MicroService(), app.m)
+        mock.m.get_wrapper.assert_called_with()
+        self.assertEqual(mock.m.get_wrapper(), app.wrapper)
+
+    def test_main(self):
+        import_app("__main__")
+        mock.wrapper.Base.metadata.create_all.assert_called_with(bind=mock.wrapper.engine)
+        mock.m.run.assert_called_with()
+
+    def test_not_main(self):
```
Comment (use-to) [inline app/tests/test_app.py]: Not meaningful test names. Rethink about this naming.

File: app/tests/test_app.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        import_app()
+        mock.wrapper.Base.metadata.create_all.assert_not_called()
+        mock.m.run.assert_not_called()
+
+    def test_endpoints(self):
```
Comment (use-to) [inline app/tests/test_app.py]: Here too.

File: app/tests/test_app.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        app = import_app("__main__")
+        elements = [getattr(app, element_name) for element_name in dir(app)]
+
+        for path, requires in [
```
Comment (use-to) [inline app/tests/test_app.py]: Define a variable for this list to improve readability.

File: app/tests/test_app.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            (["inventory", "list"], {}),
+            (["shop", "list"], {}),
+            (["shop", "info"], shop_info),
+            (["shop", "buy"], shop_buy),
+        ]:
+            self.assertIn((path, requires), mock.user_endpoints)
+            self.assertIn(mock.user_endpoint_handlers[tuple(path)], elements)
+
+        for path in [
```
Comment (use-to) [inline app/tests/test_app.py]: Outsource this list, too.

File: app/tests/test_inventory.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class TestInventory(TestCase):
+    def setUp(self):
+        mock.reset_mocks()
+
+        self.query_inventory = mock.MagicMock()
+        mock.wrapper.session.query.side_effect = {inventory.Inventory: self.query_inventory}.__getitem__
+
+    def test_list_inventory(self):
```
Comment (use-to) [inline app/tests/test_inventory.py]: This isnt really explicit.
We have to write a testnaming guide.

File: app/tests/test_inventory.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        elements = [mock.MagicMock() for _ in range(5)]
+        self.query_inventory.filter_by.return_value = elements
+        self.assertEqual({"elements": [e.serialize for e in elements]}, inventory.list_inventory({}, "the-user"))
```
Comment (use-to) [inline app/tests/test_inventory.py]: Just reduce complexity here.

File: app/tests/test_inventory.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        self.query_inventory.filter_by.assert_called_with(owner="the-user")
+
+    def test_exists_not_found(self):
```
Comment (use-to) [inline app/tests/test_inventory.py]: I think the prefix `test` is unnecessary -> Testing Guide

File: app/tests/test_inventory.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        self.query_inventory.filter_by().first.return_value = None
+        self.assertEqual({"exists": False}, inventory.exists({"owner": "the-user", "item_name": "my-item"}, ""))
```
Comment (use-to) [inline app/tests/test_inventory.py]: Reduce complexity of statements.

File: app/tests/test_inventory.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        self.query_inventory.filter_by.assert_called_with(owner="the-user", element_name="my-item")
+        self.query_inventory.filter_by().first.assert_called_with()
```
Comment (use-to) [inline app/tests/test_inventory.py]: Is it possible to merge this line with the line above? Like:
```suggestion
        self.query_inventory.filter_by(owner="the-user", element_name="my-item").first.assert_called_with()
```

File: app/tests/test_inventory.py (hunk 8 above)
Comment (Defelo) [reply app/tests/test_inventory.py]: > The individual tests are defined with methods whose names start with the letters `test`.

No, the prefix is necessary

File: app/tests/test_inventory.py (hunk 10 above)
Comment (Defelo) [reply app/tests/test_inventory.py]: No this is not possible. The arguments of the `filter_by` call are ignored by MagicMocks

Comment (use-to) [reply app/tests/test_inventory.py]: Okay. All right, thank you!

**Comment by Defelo** [comment]
We could add the build status and the coverage to the readme. But for the coverage we would have to allow coveralls.io for the organization @use-to

**Pull request merged**